src/python/pants/backend/scala/util_rules/versions.py

```diff
@@ -3,14 +3,74 @@
 from __future__ import annotations
 
 from dataclasses import dataclass
+from enum import Enum
 
+from pants.base.deprecated import warn_or_error
 from pants.engine.rules import collect_rules, rule
 from pants.jvm.resolve.common import Coordinate
+from pants.util.strutil import softwrap
+
+
+class InvalidScalaVersion(ValueError):
+    def __init__(self, scala_version: str) -> None:
+        super().__init__(
+            softwrap(
+                f"""Value '{scala_version}' is not a valid Scala version.
+            It should be formed of [major].[minor].[patch]"""
+            )
+        )
+
+
+class ScalaCrossVersionMode(Enum):
+    PARTIAL = "partial"
+    BINARY = "binary"
+    FULL = "full"
+
+    @classmethod
+    def from_str(cls, value: str) -> ScalaCrossVersionMode:
+        if value == ScalaCrossVersionMode.PARTIAL.value:
+            warn_or_error(
+                "2.21.0",
+                f"Scala cross version: {value}",
+                "Use value `binary` instead",
+                start_version="2.20.0",
+            )
+        return cls(value)
+
+
+@dataclass(frozen=True)
+class ScalaVersion:
+    major: int
+    minor: int
+    patch: int
+
+    @classmethod
+    def parse(cls, scala_version: str) -> ScalaVersion:
+        version_parts = scala_version.split(".")
+        if len(version_parts) != 3:
+            raise InvalidScalaVersion(scala_version)
+        return cls(
+            major=int(version_parts[0]), minor=int(version_parts[1]), patch=int(version_parts[2])
```

Comment:
I'm not very familiar with JVM versioning, but this would fail on any characters like `rc1` or `.alpha`. Do you know if that's a risk?

Reply:
A quick scan of Maven repositories reveals that yeah, that could be a risk. At the end of the version is possible to have an arbitrary string separated by a `-` for non final versions.

I'll switch this to a regex match.